### PR TITLE
[ci] use GitHub Actions to re-generate R configure

### LINKED ISSUE #3283: [R-package] automate release process for R
See https://github.com/microsoft/LightGBM/pull/3071#issuecomment-670243598 for background. To close this issue, implement the solution in that comment or some other approach that allows for building a release of the R package without manually updating anything other than `VERSION.txt`

### PR DESCRIPTION
**Pull request by StrikerRUS**
Closed #3283.

Refer to https://github.com/microsoft/LightGBM/pull/4117#pullrequestreview-622620015 for the initial discussion.

Refer to https://github.com/StrikerRUS/LightGBM/pull/1 for the usage example.

~Push from a workflow won't trigger our regular workflows:~
> ~For example, if a workflow run pushes code using the repository's `GITHUB_TOKEN`, a new workflow will not run even when the repository contains a workflow configured to run when `push` events occur.~
  ~https://docs.github.com/en/actions/reference/events-that-trigger-workflows#triggering-new-workflows-using-a-personal-access-token~

~I don't think that this a problem for us, because anyway we need a follow-up commit with reverting back to `~~VERSION~~` in `configure.ac`, right?~

### CODE REVIEW

File: .github/workflows/r_configure.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+          ref: "refs/heads/${{ fromJSON(github.event.client_payload.pr_branch) }}"
+      - name: Update configure
+        shell: bash
+        run: ./R-package/recreate-configure.sh || exit -1
+      - name: Push changes
+        run: |
+          git config --global user.name "GitHub Actions Bot"
+          git config --global user.email "githubactionsbot@users.noreply.github.com"
+          git add "./R-package/configure"
+          git commit --allow-empty -m"Auto-update configure"
```
Comment (jameslamb) [inline .github/workflows/r_configure.yml:37]:
```suggestion
          git commit --allow-empty -m "Auto-update configure"
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
>  because anyway we need a follow-up commit with reverting back to ~~VERSION~~ in configure.ac, right?

No, no followup commit is required because there is never a commit that removes the `~~VERSION~~` in `configure.ac`. I'm confused, how is `configure.ac` getting updated and showing up in the diff for https://github.com/StrikerRUS/LightGBM/pull/2? Looking at the workflow template, I don't understand how `configure.ac` is getting changed OR how that change is getting committed.

Notice it was unchanged in the diff for the most recent release: https://github.com/microsoft/LightGBM/pull/3872/files, and `recreate-configure.sh` does that templating with a **temporary** `configure.ac`: https://github.com/microsoft/LightGBM/blob/f879018b50c2b323ca0409fb5cc338e67ff648f4/R-package/recreate-configure.sh#L25-L26.

**Comment by StrikerRUS** [comment]
> I'm confused, how is `configure.ac` getting updated and showing up in the diff for StrikerRUS#2?

Just like it is suggested in https://github.com/microsoft/LightGBM/tree/master/R-package#changing-the-cran-package (**Linux or Mac** section):
1. edit `configure.ac`
2. comment `/gha run r-configure`
3. revert step `#1`

How this process should be done right?

**Comment by jameslamb** [comment]
"Edit `configure.ac`" there is not about changing the version. It's about making *functional* changes, for example like the one suggested in #4131.

If the only thing that is changing is the version of LightGBM, you do not need to touch `configure.ac` at all.

**Comment by StrikerRUS** [comment]
> If the only thing that is changing is the version of LightGBM, you do not need to touch `configure.ac` at all.

OK, got it. What should be touched then?

**Comment by jameslamb** [comment]
> > If the only thing that is changing is the version of LightGBM, you do not need to touch `configure.ac` at all.
> 
> OK, got it. What should be touched then?

you don't need to manually edit any files. Just running `recreate-configure.sh` will create a new `configure` script.

**Comment by StrikerRUS** [comment]
> you don't need to manually edit any files.

Ah, seems I'm getting what's going on! Just manually edit `VERSION.txt`, right?

**Comment by jameslamb** [comment]
Yep! Just manually edit `VERSION.txt` if that's changing. Which I think is ok.

Sorry if the instructions weren't clear. When I wrote https://github.com/microsoft/LightGBM/tree/master/R-package#changing-the-cran-package , the main audience I had in mind was contributors making changes like #4131, not our usual maintenance like doing a new release. If there are changes to the language there that you think would make it more clear, I'd be happy to review those as part of this PR.

**Comment by StrikerRUS** [comment]
OK, @jameslamb please take one more look. I hope I understood it right this time.
Here is new demo: https://github.com/StrikerRUS/LightGBM/pull/1.

File: R-package/README.md [hunk 2]
```diff
@@ -308,6 +308,10 @@ At build time, `configure` will be run and used to create a file `Makevars`, usi
 
 3. Edit `src/Makevars.in`.
 
+Alternatively, GitHub Actions can re-generate this file for you. On a pull request (only on internal one, not works for ones from forks), create a comment with this phrase:
```
Comment (jameslamb) [inline R-package/README.md:311]: small suggested change to this language

```suggestion
Alternatively, GitHub Actions can re-generate this file for you. On a pull request (only on internal one, does not work for ones from forks), create a comment with this phrase:
```

File: .github/workflows/r_configure.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+          token: ${{ secrets.WORKFLOW }}
+          persist-credentials: true
+      - name: Update configure
+        shell: bash
+        run: ./R-package/recreate-configure.sh || exit -1
+      - name: Push changes
+        run: |
+          git config --global user.name "GitHub Actions Bot"
+          git config --global user.email "githubactionsbot@users.noreply.github.com"
+          git add "./R-package/configure"
```
Comment (jameslamb) [inline .github/workflows/r_configure.yml:36]: I can tell your main OS is Windows because of your defensive use of quotes :stuck_out_tongue:

**Review by jameslamb — APPROVED** [review]
This looks great, thanks! The diff in https://github.com/StrikerRUS/LightGBM/pull/1/files#diff-dadfc37f082f218f0e2f0adfc1affc7c55044bcdf8dff110aab9cdfda842d6f3 looks exactly like what I'd expect. Just one minor grammar comment, but I think the workflow is perfect.

I think it is ok that the commit from this bot won't trigger a new CI build. Usually on release PRs (the main place where we update `configure`) there are additional commits after updating `configure` that don't have any impact on `configure`.

For cases like #4131 , this will have to be something that we "just know" to re-run if the contributor doesn't manually regenerate `configure`. I don't think that's problematic at all, as that file very rarely receives updates that are not part of the release process: https://github.com/microsoft/LightGBM/commits/master/R-package/configure.

**Comment by StrikerRUS** [comment]
Thanks!

> looks exactly like what I'd expect. 

Great!

> I think it is ok that the commit from this bot won't trigger a new CI build

It **does** with the latest code! 🙂 Thanks to `token` and `persist-credentials` arguments of `actions/checkout`.


image


> For cases like #4131 , ...

Yeah, unfortunately, GitHub Actions cannot push into forked repos. So we can't do anything with it.

**Pull request merged**